### ISSUE TITLE
Fix issues found by schema validator

### ISSUE THREAD

**Issue by CelticMinstrel** [wesnoth/wesnoth#2779]
I've been working on the schema validator (#2562) so that we can track down and fix hard-to-find bugs. This issue stands to get feedback on some cases where I'm not sure what the correct path is, as well as to serve as a checklist for the validation.

For completeness, I intend to validate every campaign on every difficulty, as well as MP, the editor, and tests. To be clear, all I actually need to do is load each one up, look at the log, fix any errors, and repeat until there are no errors - I'm not going to be playing through any of the campaigns or anything. I don't expect to find any errors on medium or hard that aren't also on easy, so I'll probably just load medium and hard (and nightmare where applicable) once just to be sure.

* [x] Core WML (that which is loaded by simply getting to the title screen)
* [x] Editor WML
* [x] Interactive tests (accessible by hotkey)
* [x] Unit tests (not accessible by hotkey)
* [x] Multiplayer WML (that which is loaded by simply getting to the create screen)
* [x] Hornshark Island (2P)
* [x] A New Land (4P)
* [x] Team Survival (6P)
* [x] Legend of Wesmere MP
* [x] Tutorial
* [x] An Orcish Incursion
* [x] Dead Water
* [x] Defador's Memoirs
* [x] Descent into Darkness
* [x] Eastern Invasion
* [x] Hammer of Thursagan
* [x] Heir to the Throne
* [x] Legend of Wesmere
* [x] Liberty
* [x] Northern Rebirth
* [x] Rise of Wesnoth
* [x] Sceptre of Fire
* [x] Secrets of the Ancients
* [x] Son of the Black Eye
* [x] South Guard
* [x] Two Brothers
* [x] Under the Burning Suns
* [ ] Under the Burning Suns (old units) - I know they're going to be removed soon, but I figure checking them anyway could potentially help find bugs in the schema itself

**Comment by CelticMinstrel** [comment]
Alright, getting some errors in Legend of Wesmere that I'm not sure how to deal with. Here's a cherry-picked log of the issues (note that the line numbers may be slightly off since I already made some edits in LoW):

```
20180330 22:03:14 error validation: Invalid value 'stage[main_loop].candidate_action' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:48
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/01_The_Uprooting.cfg:260
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'aspect[recruitment_instructions].facet' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:15
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:344
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'aspect[avoid].facet' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:15
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:275
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:345
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'goal' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:37
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:288
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:356
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'goal' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:37
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:300
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:367
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'goal' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:37
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:288
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:378
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'goal' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:37
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:300
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:379
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'aspect[avoid].facet' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:15
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:275
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:390
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
20180330 22:03:14 error validation: Invalid value 'goal' in key 'path=' in tag [modify_ai]
at core/macros/ai.cfg:37
    included from core/_main.cfg:18
    included from _main.cfg:18
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:312
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
    included from campaigns/Legend_of_Wesmere/scenarios/chapter1/02_Hostile_Mountains.cfg:391
    included from campaigns/Legend_of_Wesmere/_main.cfg:15
    included from _main.cfg:38
```

And that's just in the first two scenarios; there are tons more like that. The key that's causing the errors is the lack of any `[]` on the end of the path, but I'm not sure whether that should be accepted (and I change the schema) or if it should be _not_ accepted (and I change LoW). Any opinions?

Even if it's acceptable it would only be for `action=add`... which is what they're being used for here, anyway.

**Comment by CelticMinstrel** [comment]
I have a question about `[endlevel]` (maybe @gfgtdf would know?). The wiki claims (well, strongly implies) that with `result=defeat` all other keys are ignored. Based on this I had set the schema so that these keys are only allowed with `result=victory`. The Lua code doesn't particularly imply this, but it mostly forwards the arguments to C++, so I'm not sure how it should be. Legend of Wesmere regularly uses the full set of keys even without `result=victory`, but that doesn't necessarily mean those keys actually do anything.

So basically, is there any point in using the additional keys with `result=defeat`? Or maybe some of them will have an effect while others won't? Or, is the wiki simply wrong on this count?

**Comment by gfgtdf** [comment]
> I have a question about [endlevel] (maybe @gfgtdf would know?). The wiki claims (well, strongly implies) that with result=defeat all other keys are ignored

I am not sure about the 1.12 behviour but in 1.14 it si actualyl possibel to get a defeat dialog and still continue to the next scenario in which case most of the keys still work. This was made to support vs mp campaigns where one side wins, and the other looses and the campaign goes on still (liek the last chapter of LoW). 

It could probably be used in sp aswell to have soem scenario where you are supposed to loose? I didn't really think about that case when i implement it.


This code https://github.com/wesnoth/wesnoth/blob/1.13.13/src/game_initialization/playcampaign.cpp#L120 is also relevant which generates the carroyver message.

**Comment by CelticMinstrel** [comment]
I have a few remaining errors in HTTT that I'm not quite sure what to do with...

```
20180404 22:11:56 error validation: Tag [story] may not be used in [root]
at campaigns/Heir_To_The_Throne/utils/intro.cfg:2
    included from campaigns/Heir_To_The_Throne/_main.cfg:88
    included from _main.cfg:38
20180404 22:12:01 error validation: Invalid key 'config_name=' in tag [generator]
at campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg:9
    included from campaigns/Heir_To_The_Throne/_main.cfg:94
    included from _main.cfg:38
20180404 22:12:05 error validation: Invalid key 'bonus=' in tag [scenario]
at campaigns/Heir_To_The_Throne/scenarios/22_Return_to_Wesnoth.cfg:9
    included from campaigns/Heir_To_The_Throne/_main.cfg:94
    included from _main.cfg:38
```

**Comment by CelticMinstrel** [comment]
I also have a few remaining issues in SoF, which are basically similar to those in HTTT:

```
20180405 21:26:47 error validation: Invalid key 'config_name=' in tag [generator]
at campaigns/Sceptre_of_Fire/scenarios/4_Gathering_Materials.cfg:21
    included from campaigns/Sceptre_of_Fire/_main.cfg:117
    included from _main.cfg:38
20180405 21:26:48 error validation: Invalid key 'village_gold=' in tag [scenario]
at campaigns/Sceptre_of_Fire/scenarios/7_Outriding_the_Outriders.cfg:9
    included from campaigns/Sceptre_of_Fire/_main.cfg:117
    included from _main.cfg:38
```

**Comment by CelticMinstrel** [comment]
@Vultraz pointed out that `config_name` is indeed valid while toplevel `[story]` is not, but I still need to know whether scenario-level `bonus` and `village_gold` are allowed.

**Comment by CelticMinstrel** [comment]
The errors in TRoW are mostly more of the same, but there's one different one:

```
20180405 23:37:03 error validation: Invalid key 'image=' in tag [side]
at campaigns/The_Rise_Of_Wesnoth/scenarios/16_The_Kalian.cfg:49
    included from campaigns/The_Rise_Of_Wesnoth/_main.cfg:94
    included from _main.cfg:38
```

Is it allowed to override a unit's base image in `[unit]`? Even if it were I wouldn't expect it to work as you'd expect?
